### PR TITLE
[Form] Take group into account when guessing if property is required

### PR DESCRIPTION
**Pull request by renanbr**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | https://github.com/EasyCorp/EasyAdminBundle/issues/2673
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php [hunk 1]
```diff
@@ -162,13 +162,21 @@ public function guessTypeForConstraint(Constraint $constraint)
      *
      * @return ValueGuess|null The guess whether the field is required
      */
-    public function guessRequiredForConstraint(Constraint $constraint)
+    public function guessRequiredForConstraint(Constraint $constraint, $class = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php]: this change is a BC break...

File: src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php [hunk 2]
```diff
@@ -168,7 +168,16 @@ public function guessRequiredForConstraint(Constraint $constraint)
             case 'Symfony\Component\Validator\Constraints\NotNull':
             case 'Symfony\Component\Validator\Constraints\NotBlank':
             case 'Symfony\Component\Validator\Constraints\IsTrue':
-                return new ValueGuess(true, Guess::HIGH_CONFIDENCE);
+                $groups = ['Default'];
+
+                if (\func_num_args() > 1 && $class = func_get_arg(1)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php]: We should now document this somehow. Do we want to add the argument to the signature on v6?

File: src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php [hunk 3]
```diff
@@ -263,7 +263,10 @@ protected function guess($class, $property, \Closure $closure, $defaultValue = n
         if ($classMetadata instanceof ClassMetadataInterface && $classMetadata->hasPropertyMetadata($property)) {
             foreach ($classMetadata->getPropertyMetadata($property) as $memberMetadata) {
                 foreach ($memberMetadata->getConstraints() as $constraint) {
-                    if ($guess = $closure($constraint)) {
+                    $groups = [Constraint::DEFAULT_GROUP];
+                    $classParts = explode('\\', $class);
+                    $groups[] = end($classParts);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php]: `$groups[] = substr(strrstr($class, '\\'), 1) ?: $class;`?

File: src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php [hunk 4]
```diff
@@ -84,6 +84,26 @@ public function testGuessRequired($constraint, $guess)
         $this->assertEquals($guess, $this->guesser->guessRequired(self::TEST_CLASS, self::TEST_PROPERTY));
     }
 
+    public function guessRequiredProviderWithGroup()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php]: I think we prefer putting data providers after the corresponding test cases

**Review by nicolas-grekas — COMMENTED** [review]
Just minor comments - the implementation looks much simpler now :)

File: src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php (hunk 3 above)
Comment (renanbr) [reply src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php]: done via 90e4d04b856f39ce831188116f1fd2731029a0c2

File: src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php (hunk 4 above)
Comment (renanbr) [reply src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php]: done via 63959fdc43e527c477deda9f36c67c7b3e23379d

**Comment by xabbuh** [comment]
I am afraid that these changes would break cases where someone puts constraints in a different group than the default one, but also validates the form in this group. Currently, the constraints would be applied while that wouldn't be true anymore after this.

To account for this we would have to look for the value of the `validation_groups` option of the root form and take the configured value into account too. But that's way more complicated.

Status: Needs work

**Comment by fabpot** [comment]
@renanbr What do you think if the last @xabbuh comment? Still worth implementing? Or shall we close?

**Pull request closed without merging**